npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb

```diff
@@ -2,18 +2,213 @@
+      def initialize(raw_version)
+        super(
+          NAME,
+          Version.new(raw_version),
+          DEPRECATED_VERSIONS,
+          SUPPORTED_VERSIONS
+        )
+      end
+
+      sig { override.returns(T::Boolean) }
+      def deprecated?
+        false
+      end
+
+      sig { override.returns(T::Boolean) }
+      def unsupported?
+        false
+      end
+    end
+
+    class YarnPackageManager < Ecosystem::VersionManager
+      extend T::Sig
+      NAME = "yarn"
+      LOCKFILE_NAME = "yarn.lock"
+
+      YARN_V1 = "1"
+      YARN_V2 = "2"
+      YARN_V3 = "3"
+
+      SUPPORTED_VERSIONS = T.let([
+        Version.new(YARN_V1),
+        Version.new(YARN_V2),
+        Version.new(YARN_V3)
+      ].freeze, T::Array[Dependabot::Version])
+
+      DEPRECATED_VERSIONS = T.let([].freeze, T::Array[Dependabot::Version])
+
+      sig { params(raw_version: String).void }
+      def initialize(raw_version)
+        super(
+          NAME,
+          Version.new(raw_version),
+          DEPRECATED_VERSIONS,
+          SUPPORTED_VERSIONS
+        )
+      end
+
+      sig { override.returns(T::Boolean) }
+      def deprecated?
+        false
+      end
+
+      sig { override.returns(T::Boolean) }
+      def unsupported?
+        false
+      end
+    end
+
+    class PNPMPackageManager < Ecosystem::VersionManager
+      extend T::Sig
+      NAME = "pnpm"
+      LOCKFILE_NAME = "pnpm-lock.yaml"
+
+      PNPM_V7 = "7"
+      PNPM_V8 = "8"
+      PNPM_V9 = "9"
+
+      SUPPORTED_VERSIONS = T.let([
+        Version.new(PNPM_V7),
+        Version.new(PNPM_V8),
+        Version.new(PNPM_V9)
+      ].freeze, T::Array[Dependabot::Version])
+
+      DEPRECATED_VERSIONS = T.let([].freeze, T::Array[Dependabot::Version])
+
+      sig { params(raw_version: String).void }
+      def initialize(raw_version)
+        super(
+          NAME,
+          Version.new(raw_version),
+          DEPRECATED_VERSIONS,
+          SUPPORTED_VERSIONS
+        )
+      end
+
+      sig { override.returns(T::Boolean) }
+      def deprecated?
+        false
+      end
+
+      sig { override.returns(T::Boolean) }
+      def unsupported?
+        false
+      end
+    end
+
+    PACKAGE_MANAGER_CLASSES = {
+      NpmPackageManager::NAME => NpmPackageManager,
+      YarnPackageManager::NAME => YarnPackageManager,
+      PNPMPackageManager::NAME => PNPMPackageManager
+    }.freeze
+
+    class PackageManagerDetector
+      extend T::Sig
+      extend T::Helpers
+
+      sig do
+        params(
+          lockfiles: T::Hash[Symbol, T.nilable(Dependabot::DependencyFile)],
+          package_json: T::Hash[String, T.untyped]
+        ).void
+      end
+      def initialize(lockfiles, package_json)
+        @lockfiles = lockfiles
+        @package_json = package_json
+        @manifest_package_manager = package_json["packageManager"]
+        @engines = package_json.fetch("engines", nil)
+      end
+
+      sig { returns(T.nilable(String)) }
```

Comment:
can this return a strongly-typed value to minimize the risk of typos?

Reply:
other then npm all others are replaced to be `String` can't be nil. Also current flow doesn't allow other then `npm`, `yarn`, and `pnpm` to return here. Since we have `npm` as default in other cases then this package manager we are returning the default `npm`.